internal/controller/grafanaorganization_controller.go

```diff
@@ -39,24 +44,91 @@ type GrafanaOrganizationReconciler struct {
 
 // Reconcile is part of the main Kubernetes reconciliation loop which aims to
 // move the current state of the GrafanaOrganization closer to the desired state.
-// TODO(zirko): Modify the Reconcile function to compare the state specified by
-// the GrafanaOrganization object against the actual organization state, and then
-// perform operations to make the organization state reflect the state specified by
-// the user.
 //
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.16.0/pkg/reconcile
 func (r *GrafanaOrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := log.FromContext(ctx)
 
-	// TODO(zirko): your logic here
+	logger.Info("Started reconciling Grafana Organization")
+	defer logger.Info("Finished reconciling Grafana Organization")
+
+	grafanaOrganization := &v1alpha1.GrafanaOrganization{}
+	err := r.Client.Get(ctx, req.NamespacedName, grafanaOrganization)
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(client.IgnoreNotFound(err))
+	}
+
+	logger.WithValues("grafanaOrganization", grafanaOrganization.ObjectMeta.Name)
+
+	// Generate Grafana client
+	grafanaAPI, err := grafanaClient.GenerateGrafanaClient(ctx, r.Client, logger)
+	if err != nil {
+		logger.Error(err, "Failed to create Grafana admin client")
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	// Test connection to Grafana
+	_, err = grafanaAPI.Health.GetHealth()
+	if err != nil {
+		logger.Error(err, "Failed to connect to Grafana")
+		return ctrl.Result{}, errors.WithStack(err)
+	}
+
+	logger.Info("Successfully connected to Grafana, lets start hacking...")
+
+	// Handle deleted grafana organizations
+	if !grafanaOrganization.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, r.reconcileDelete(ctx, grafanaOrganization)
+	}
+
+	// Handle non-deleted grafana organizations
+	return r.reconcileCreate(ctx, grafanaAPI, grafanaOrganization)
+}
+
+// reconcileCreate creates the grafanaOrganization.
+func (r GrafanaOrganizationReconciler) reconcileCreate(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, grafanaOrganization *v1alpha1.GrafanaOrganization) (ctrl.Result, error) { // nolint:unparam
```

Comment:
I don't think we need the `nolint` marker anymore.

```suggestion
func (r GrafanaOrganizationReconciler) reconcileCreate(ctx context.Context, grafanaAPI *grafanaAPI.GrafanaHTTPAPI, grafanaOrganization *v1alpha1.GrafanaOrganization) (ctrl.Result, error) {
```

Reply:
Sadly we do for now, but we should remove it in the next PRs

Reply:
It looks like all the params are used already to me. But yeah, it can be removed later.

Reply:
The issue is not about params but returned value, we apparently always return nil or empty values for ctrl.Result

Reply:
oooh ok I see